src/ripple/ledger/impl/CashDiff.cpp

```diff
@@ -0,0 +1,731 @@
+}
+
+bool CashDiff::rmDust()
+{
+    return impl_->rmDust();
+}
+
+bool CashDiff::rmLhsDeletedOffers()
+{
+    return impl_->rmLhsDeletedOffers();
+}
+
+bool CashDiff::rmRhsDeletedOffers()
+{
+    return impl_->rmRhsDeletedOffers();
+}
+
+//------------------------------------------------------------------------------
+
+// Function that compares two STAmounts and returns true if the difference
+// is dust-sized.
+bool diffIsDust (STAmount const& v1, STAmount const& v2, std::uint8_t e10)
+{
+    // If one value is positive and the other negative then there's something
+    // odd afoot.
+    if (v1 != zero && v2 != zero && (v1.negative() ^ v2.negative()))
+        return false;
+
+    // v1 and v2 must be the same Issue for their difference to make sense.
+    if (v1.native() ^ v2.native())
+        return false;
+
+    if (!v1.native() && (v1.issue() != v2.issue()))
+        return false;
+
+    // If v1 == v2 then the dust is vanishingly small.
+    if (v1 == v2)
+        return true;
+
+    STAmount const& small = v1 < v2 ? v1 : v2;
+    STAmount const& large = v1 < v2 ? v2 : v1;
+
+    // Handling XRP is different from IOU.
+    if (v1.native())
+    {
+        std::uint64_t const s = small.mantissa();
+        std::uint64_t const l = large.mantissa();
+
+        // Always allow a couple of drops of noise.
+        if (l - s <= 2)
+            return true;
+
+        static_assert (sizeof (1ULL) == sizeof (std::uint64_t), "");
+        std::uint64_t ratio = s / (l - s);
+        static constexpr std::uint64_t e10Lookup[]
+        {
+                                     1ULL,
+                                    10ULL,
+                                   100ULL,
+                                 1'000ULL,
+                                10'000ULL,
+                               100'000ULL,
+                             1'000'000ULL,
+                            10'000'000ULL,
+                           100'000'000ULL,
+                         1'000'000'000ULL,
+                        10'000'000'000ULL,
+                       100'000'000'000ULL,
+                     1'000'000'000'000ULL,
+                    10'000'000'000'000ULL,
+                   100'000'000'000'000ULL,
+                 1'000'000'000'000'000ULL,
+                10'000'000'000'000'000ULL,
+               100'000'000'000'000'000ULL,
+             1'000'000'000'000'000'000ULL,
+            10'000'000'000'000'000'000ULL,
+        };
+        static std::size_t constexpr maxIndex =
+            sizeof (e10Lookup) / sizeof e10Lookup[0];
+
+        // Make sure the table is big enough.
+        static_assert (e10Lookup[maxIndex - 1] * 10U < e10Lookup[maxIndex - 1],
+            "Table too small");
+
+        if (e10 >= maxIndex)
+            return false;
+
+        return ratio >= e10Lookup[e10];
+    }
+
+    // Non-native.  Note that even though large and small may not be equal,
+    // their difference may be zero if they are large values.  We treat
```

Comment:
I think it's when their mantissas differ by less than ten, not when they are "large values".